File: scripts/test_account_selection.py
#!/usr/bin/env python3
"""浏览器测试：新建任务弹窗 account_ids 勾选逻辑"""
from __future__ import annotations

import json
import sys
from playwright.sync_api import sync_playwright

BASE = "http://127.0.0.1:8765"


def get_checked_ids(page):
    return page.eval_on_selector_all(
        "#taskDialog .account-groups input:checked",
        "els => els.map(el => el.dataset.account)",
    )


def main() -> int:
    captured: dict | None = None
    fails = 0

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()

        def on_request(request):
            nonlocal captured
            if request.url.endswith("/api/tasks") and request.method == "POST":
                captured = json.loads(request.post_data or "{}")

        page.on("request", on_request)
        page.goto(BASE, wait_until="networkidle")
        page.click("#newTaskBtn")
        page.wait_for_selector("#dialogQqAccounts input")

        all_ids = page.eval_on_selector_all(
            "#taskDialog .account-groups input",
            "els => els.map(el => el.dataset.account)",
        )
        qq_ids = page.eval_on_selector_all(
            "#dialogQqAccounts input", "els => els.map(el => el.dataset.account)"
        )
        bot_ids = page.eval_on_selector_all(
            "#dialogBotAccounts input", "els => els.map(el => el.dataset.account)"
        )
        print(f"accounts total={len(all_ids)} qq={len(qq_ids)} bot={len(bot_ids)}")

        checked = get_checked_ids(page)
        ok = len(checked) == len(all_ids)
        print(f"TEST default all checked: {'PASS' if ok else 'FAIL'} ({len(checked)}/{len(all_ids)})")
        fails += 0 if ok else 1

        page.click("#dialogDeselectQq")
        checked = get_checked_ids(page)
        bot_only = (
            len(checked) == len(bot_ids)
            and all(i in checked for i in bot_ids)
            and all(i not in checked for i in qq_ids)
        )
        print(f"TEST deselect QQ keeps bot: {'PASS' if bot_only else 'FAIL'} {checked}")
        fails += 0 if bot_only else 1

        page.click("#dialogSelectAllCh")
        page.evaluate(
            """() => {
            dialog.videos = [{ id: 'test123', title: '测试视频', link: '', play_addr: '', pic: '', author: '', platform: 'bili' }];
            dialog.selectedVideos = new Set(['test123']);
            renderDialogVideos();
            updateDialogBtns();
        }"""
        )

        captured = None
        page.click("#dialogCreate")
        page.wait_for_timeout(1000)

        if not captured:
            print("TEST submit payload bot-only: FAIL (no POST captured)")
            fails += 1
        else:
            sent = captured.get("account_ids") or []
            ok = (
                len(sent) == len(bot_ids)
                and all(i in sent for i in bot_ids)
                and all(i not in sent for i in qq_ids)
            )
            print(f"TEST submit account_ids bot-only: {'PASS' if ok else 'FAIL'} {sent}")
            fails += 0 if ok else 1

        page.click("#newTaskBtn")
        page.wait_for_selector("#dialogQqAccounts input")
        page.click("#dialogDeselectQq")
        page.click("#dialogDeselectBot")
        checked = get_checked_ids(page)
        print(f"TEST all unchecked count: {len(checked)}")

        page.evaluate(
            """() => {
            dialog.videos = [{ id: 'test456', title: '测试2', link: '', platform: 'bili' }];
            dialog.selectedVideos = new Set(['test456']);
            const ch = state.channels[0];
            if (ch) dialog.selectedChannels = new Set([`${ch.guild_id}:${ch.channel_id}`]);
            updateDialogBtns();
        }"""
        )

        captured = None
        page.on("dialog", lambda d: d.accept())
        page.click("#dialogCreate")
        page.wait_for_timeout(800)

        if captured:
            print(f"TEST empty selection blocked: FAIL (still posted) {captured.get('account_ids')}")
            fails += 1
        else:
            print("TEST empty selection blocked: PASS (no POST)")

        browser.close()

    return fails


if __name__ == "__main__":
    sys.exit(main())

File: scripts/test_account_selection.mjs
/**
 * 浏览器测试：新建任务弹窗 account_ids 勾选逻辑
 * 用法: node scripts/test_account_selection.mjs
 */
import { chromium } from "playwright";

const BASE = process.env.BASE_URL || "http://127.0.0.1:8765";

function getCheckedIds(page) {
  return page.$$eval("#taskDialog .account-groups input:checked", (els) =>
    els.map((el) => el.dataset.account)
  );
}

async function main() {
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();

  let capturedBody = null;
  await page.route("**/api/tasks", async (route, request) => {
    if (request.method() === "POST") {
      capturedBody = JSON.parse(request.postData() || "{}");
    }
    await route.continue();
  });

  await page.goto(BASE, { waitUntil: "networkidle" });
  await page.click("#newTaskBtn");
  await page.waitForSelector("#dialogQqAccounts input");

  const allIds = await page.$$eval("#taskDialog .account-groups input", (els) =>
    els.map((el) => el.dataset.account)
  );
  const qqIds = await page.$$eval("#dialogQqAccounts input", (els) =>
    els.map((el) => el.dataset.account)
  );
  const botIds = await page.$$eval("#dialogBotAccounts input", (els) =>
    els.map((el) => el.dataset.account)
  );

  console.log("accounts total", allIds.length, "qq", qqIds.length, "bot", botIds.length);

  // 默认全选
  let checked = await getCheckedIds(page);
  console.log("TEST default all checked:", checked.length === allIds.length ? "PASS" : "FAIL", checked.length, allIds.length);

  // 清空 QQ，只留 Bot
  await page.click("#dialogDeselectQq");
  checked = await getCheckedIds(page);
  const expectBotOnly = botIds.every((id) => checked.includes(id)) && qqIds.every((id) => !checked.includes(id));
  console.log("TEST deselect QQ keeps bot:", expectBotOnly ? "PASS" : "FAIL", checked);

  // 模拟最小可提交：选频道 + 注入一条假视频
  await page.evaluate(() => {
    window.__injectVideo = true;
  });
  await page.click("#dialogSelectAllCh");

  // 注入 dialog 状态（绕过搜索）
  await page.evaluate(() => {
    dialog.videos = [{ id: "test123", title: "测试视频", link: "", play_addr: "", pic: "", author: "", platform: "bili" }];
    dialog.selectedVideos = new Set(["test123"]);
    renderDialogVideos();
    updateDialogBtns();
  });

  capturedBody = null;
  await page.click("#dialogCreate");
  await page.waitForTimeout(800);

  if (!capturedBody) {
    console.log("TEST submit payload: FAIL (no POST captured)");
  } else {
    const sent = capturedBody.account_ids || [];
    const botOnlyOk =
      sent.length === botIds.length &&
      botIds.every((id) => sent.includes(id)) &&
      qqIds.every((id) => !sent.includes(id));
    console.log("TEST submit account_ids bot-only:", botOnlyOk ? "PASS" : "FAIL", sent);
  }

  // 全不选应被拦截
  await page.click("#newTaskBtn");
  await page.waitForSelector("#dialogQqAccounts input");
  await page.click("#dialogDeselectQq");
  await page.click("#dialogDeselectBot");
  checked = await getCheckedIds(page);
  console.log("TEST all unchecked count:", checked.length);

  await page.evaluate(() => {
    dialog.videos = [{ id: "test456", title: "测试2", link: "", platform: "bili" }];
    dialog.selectedVideos = new Set(["test456"]);
    dialog.selectedChannels = new Set(["23396421665492266:634579922"]);
    updateDialogBtns();
  });

  capturedBody = null;
  page.once("dialog", (d) => d.accept());
  await page.click("#dialogCreate");
  await page.waitForTimeout(500);

  if (capturedBody) {
    console.log("TEST empty selection blocked: FAIL (still posted)", capturedBody.account_ids);
  } else {
    console.log("TEST empty selection blocked: PASS (no POST)");
  }

  await browser.close();
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
